app/src/test/java/com/davidtomas/taskyapp/features/auth/presentation/login/LoginViewModelTest.kt

```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Test
+    fun `when login button is clicked and fields are well filled, login is performed successfully`() =
+        runTest {
+            var eventCollected: LoginUiEvent?
+            val newEmail = "newEmail"
+            val newPassword = "password"
+            coEvery { authRepository.login(any()) } returns Result.Success(AuthModelMother.mock())
+            loginViewModel.onAction(LoginAction.OnEmailChanged(newEmail))
+            loginViewModel.onAction(LoginAction.OnPasswordChanged(newPassword))
+
+            loginViewModel.uiEvent.test {
+                loginViewModel.onAction(LoginAction.OnLoginButtonClick)
+                eventCollected = awaitItem()
+                assertThat(eventCollected).isEqualTo(
+                    LoginUiEvent.ShowSnackBar(
+                        UiText.StringResource(
+                            R.string.login_success
+                        )
+                    )
+                )
+            }
+        }
+
+    @Test
+    fun `when login button is clicked and fields are incorrectly filled, login returns error`() =
+        runTest {
+            var eventCollected: LoginUiEvent?
+            val newEmail = "newEmail"
+            val newPassword = "wrongPassword"
+            coEvery { authRepository.login(any()) } returns Result.Error(DataError.Network.BAD_REQUEST)
+            loginViewModel.onAction(LoginAction.OnEmailChanged(newEmail))
+            loginViewModel.onAction(LoginAction.OnPasswordChanged(newPassword))
+
+            loginViewModel.uiEvent.test {
+                loginViewModel.onAction(LoginAction.OnLoginButtonClick)
+                eventCollected = awaitItem()
+                assertThat(eventCollected).isEqualTo(
+                    LoginUiEvent.ShowSnackBar(
+                        UiText.StringResource(
+                            R.string.error_bad_request
+                        )
+                    )
+                )
+            }
+        }
+
+    @Test
+    fun `when login button is clicked and email is empty, informative snackbar is shown`() =
+        runTest {
+            var eventCollected: LoginUiEvent?
+            val newEmail = String.EMPTY_STRING
+            val newPassword = "wrongPassword"
+            coEvery { authRepository.login(any()) } returns Result.Error(DataError.Network.BAD_REQUEST)
+            loginViewModel.onAction(LoginAction.OnEmailChanged(newEmail))
+            loginViewModel.onAction(LoginAction.OnPasswordChanged(newPassword))
+
+            loginViewModel.uiEvent.test {
+                loginViewModel.onAction(LoginAction.OnLoginButtonClick)
+                eventCollected = awaitItem()
+                assertThat(eventCollected).isEqualTo(
+                    LoginUiEvent.ShowSnackBar(
+                        UiText.StringResource(
+                            R.string.error_fields_must_be_filled
+                        )
+                    )
+                )
+            }
+        }
```

Comment:
All these tests are very similar. In your test codebase the same rules count as in the production codebase regarding code duplication and clean code. It would make sense to extract this into a helper function

Reply:
What do you think about this helper function
```
 private fun addEmailAndPasswordAndReturnBadRequestWhenLogin(email: String, password: String) {
        loginViewModel.onAction(LoginAction.OnEmailChanged(email))
        loginViewModel.onAction(LoginAction.OnPasswordChanged(password))
        coEvery { authRepository.login(any()) } returns Result.Error(DataError.Network.BAD_REQUEST)

    }
```

And use it like this
```
 @Test
    fun `when login button is clicked and email is empty, informative snackbar is shown`() =
        runTest {
            addEmailAndPasswordAndReturnBadRequestWhenLogin(
                email = String.EMPTY_STRING,
                password = "newPassword"
            )

            loginViewModel.uiEvent.test {
                val eventCollected: LoginUiEvent?
                loginViewModel.onAction(LoginAction.OnLoginButtonClick)
                eventCollected = awaitItem()
                assertThat(eventCollected).isEqualTo(
                    LoginUiEvent.ShowSnackBar(
                        UiText.StringResource(
                            R.string.error_fields_must_be_filled
                        )
                    )
                )
            }
        }
```

Reply:
Rather keep the result return value of the login call as a parameter, so you can also use this for the first function I commented on which handles the success case